### ISSUE TITLE
Reach performance parity of parser with LLVM backend

### ISSUE THREAD

**Issue by dwightguth** [kframework/kore#2942]
Right now, when I run speedscope on the invocation of `kore-match-disjunction` generated by running `krun tests/free-variable-capture.lambda --search --pattern '<k> V:K </k>'` on the k frontend definition `k-distribution/pl-tutorial/1_k/3_lambda++/lesson_2`, it reports that 70% of execution time is spent in `loadDefinitions`. Of that time, 30% of the overall total is spent parsing the kore definition. This amounts to roughly 270 milliseconds since when run without profiling enabled, the application takes 900ms total. This is rather abnormally slow, since the llvm backend is capable of parsing this definition in only 50ms.

There is no reason to imagine that it should take substantially longer to parse something in efficiently-written Haskell than it should take to parse the same code with flex and a recursive descent parser like the LLVM backend uses. I would like us to try to aim for a parser of KORE in the Haskell backend that can parse the same definition in roughly the same amount of time as the LLVM Backend.

This is going to be very necessary if we intend to reach decent performance when running `krun --search --pattern` in the llvm backend. It's not urgent, but fixing it will substantially improve the user experience for such commands, so it would be good to fix as soon as we have the time.

**Comment by dwightguth** [comment]
I have started working on this; I have some code that compiles and makes use of a completely standard implementation via happy and alex. The test suite still does not build, which is my current focus. Once the test suite has been fixed and all the relevant unit tests pass, I plan to clean it up, measure its performance on a pathological real world kore definition, and, if the performance is suitably improved, submit a draft PR.
